### PR TITLE
Support Cohorts with Resources in cache/hierarchy packages

### PR DESCRIPTION
**Pull request by gabesaba**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Update cache to support Cohorts with resources. Update hierarchy to support explicit Cohorts. Part of #79

#### Special notes for your reviewer:
commit 1 : move files and refactor parsing
commit 2: support Cohorts which provide resources in `cache`. support explicit Cohorts in `hierarchy`

Users cannot use this functionality until we wire up reconciler and webhook. In the interest of keeping PRs small, I will publish those in separate PRs.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: pkg/hierarchy/manager_test.go [hunk 1]
```diff
@@ -158,11 +166,51 @@ func TestManager(t *testing.T) {
...
+			operations: func(m M) {
+				m.AddCohort(newCohort("cohort"))
+				m.AddClusterQueue(newCq("queue"))
+				m.UpdateClusterQueueEdge("queue", "cohort")
+				m.DeleteCohort("cohort")
+			},
+			wantCohorts: sets.New("cohort"),
+			wantCqs:     sets.New("queue"),
+			wantCqEdge:  []cqEdge{{"queue", "cohort"}},
+		},
```
Comment (macsko) [inline pkg/hierarchy/manager_test.go:208]: What about testing the `cleanupCohort` method? Like: 
```go
m.AddCohort(newCohort("cohort"))
m.AddClusterQueue(newCq("queue"))
m.UpdateClusterQueueEdge("queue", "cohort")
m.DeleteCohort("cohort")
m.DeleteClusterQueue("queue")
```

File: pkg/hierarchy/manager.go [hunk 2]
```diff
@@ -53,8 +53,35 @@ func (c *Manager[CQ, C]) DeleteClusterQueue(name string) {
...
+	if !ok || cohort.childCount() == 0 {
 		return
 	}
+	implicitCohort := c.cohortFactory(name)
+	c.Cohorts[implicitCohort.GetName()] = implicitCohort
+	c.rewireChildren(cohort, implicitCohort)
+}
+
+// rewireChildren is used when we are changing a Cohort
+// from an implicit to an explcit Cohort, or vice-versa.
```
Comment (macsko) [inline pkg/hierarchy/manager.go:79]: nit: explcit -> explicit

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -397,6 +398,27 @@ func (c *Cache) DeleteClusterQueue(cq *kueue.ClusterQueue) {
...
+	c.hm.AddCohort(cohort)
+	cohort.updateCohort(apiCohort)
+}
+
+func (c *Cache) DeleteCohort(apiCohort *kueuealpha.Cohort) {
+	c.Lock()
+	defer c.Unlock()
+	c.hm.DeleteCohort(apiCohort.Name)
+
+	// if the cohort still exists after deletion, it means
```
Comment (macsko) [inline pkg/cache/cache.go:414]: nit: start with upper case

File: pkg/hierarchy/manager_test.go (hunk 1 above)
Comment (gabesaba) [reply pkg/hierarchy/manager_test.go:208]: added

File: .golangci.yaml [hunk 4]
```diff
@@ -57,6 +57,9 @@ issues:
       - staticcheck
       # TODO(#768): Drop when incrementing the API version.
       text: "SA1019: constants.QueueAnnotation is deprecated"
+    - linters:
```
Comment (gabesaba) [inline .golangci.yaml:60]: tracking here: #2965

**Pull request merged**